pages/blog/graphql-authz.mdx

```diff
@@ -0,0 +1,174 @@
+
+Querying data sources from authorization rules is powerful but it adds workload to data sources and, in most cases, the data needed to perform authorization is closely related to requested entities. There is a possibility to reduce data sources workload by performing data-dependent authorization logic in the post-execution phase.
+
+With GraphQL AuthZ it’s possible for post-execution rules to require the data, related to the rule target, that is necessary for the rule to work. For example, if the rule is attached to some Object Field it could require any information about sibling fields with their relations. Another example is if the rule is attached to Object Type it could require any fields of the object with their relations.
+
+```typescript
+// creating post-execution rule
+
+const CanReadPost = postExecRule({
+  selectionSet: '{ status author { id } }',
+})(
+  (context, fieldArgs, post, parent) =>
+    post.status === 'public' || post.author.id === context.user?.id
+);
+```
+
+This technique helps to avoid additional calls to data sources by executing authorization logic on top of the entire operation result with the related extra data being populated. Since related data is often stored in the same place it can be fetched from a data source in one call instead of making one call for authorization and one call for actual data populating.
+
+## Microservices
+
+With GraphQL AuthZ it is possible to implement a centralized gateway authorization layer as well as microservice level authorization. You can choose between storing the whole authorization schema in a holistic way in a gateway independent of GraphQL schema or let GraphQL schema manage authorization as part of itself.
+
+> “Shifting this configuration out of the gateway makes subschemas autonomous, and allows them to push their own configuration up to the gateway—enabling more sophisticated schema releases.” — [schema stitching handbook](https://github.com/gmac/schema-stitching-handbook/tree/master/stitching-directives-sdl)
+
+The `@graphql-authz/directive` package provides an importable directive that can be used to annotate types and fields within subschemas and a configuration transformer that can be used on the gateway to convert the subschema directives into explicit authorization settings.
+
+```graphql
+# using @authz directive
+
+type User {
+  id: ID!
+  email: String! @authz(rules: [IsAdmin])
+  posts: [Post!]!
+}
+
+type Post @authz(rules: [CanReadPost]) {
+  id: ID!
+  title: String!
+  body: String!
+  status: Status!
+  author: User!
+}
+
+type Query {
+  users: [User!]! @authz(rules: [IsAuthenticated])
+  post(id: ID!): Post
+}
+
+type Mutation {
+  publishPost(postId: ID!): Post! @authz(rules: [CanPublishPost])
+}
+```
+
+Directives could be used in monolith architecture as well.
+
+If directives are not applicable for some reason, plain JSON auth schema could be used to describe which rules should be executed for certain Object, Interface, Field, Query, or Mutation
```

Comment:
```suggestion
If you are not pursuing a schema-first (SDL) development flow and are more a fan of the code-first approach (which does not let you specify directives), the authorization schema can also be described as a plain JSON object. Allowing you to specify the rules that should be executed on your object types, interfaces, and fields.
```

Reply:
disagree here about the 
> code-first approach (which does not let you specify directives)

it's possible to specify directives with a code-first approach, there is an example with NestJS: https://github.com/AstrumU/graphql-authz/blob/main/examples/packages/nestjs/src/index.ts#L111

Reply:
Interesting, I haven't seen this before. When using graphql-js API for building an executable schema using code this is not possible. So nestjs is definitely doing some framework magic on top of that. Maybe we should change `(which does not let you specify directives)` to `(which does not let you specify directives without framework-specific voodoo magic)`? :D

In another library I have built, I allowed mirroring directive functionality through extension fields see https://github.com/n1ru4l/graphql-public-schema-filter#usage-instructions

Reply:
Yeah, agree, I added this :) 

https://github.com/n1ru4l/graphql-public-schema-filter is a nice tool :) yeah, in graphql-authz I also allowed putting auth configuration right to extensions